File: src/prototype/lib/TrainingSample.java
import org.json.*;
import java.util.*;
import java.util.stream.*;
import java.util.function.*;
import java.io.*;

public class TrainingSample {
	public String relation;
	public boolean positive;
	public TrainingSampleParsedSentence sentence;
	public String subject;
	public String object;
	public List<Integer> subjectTokenIndices;
	public List<Integer> objectTokenIndices;

	private static String RELATION = "relation";
	private static String POSITIVE = "positive";
	private static String SENTENCE = "sentence";
	private static String SUBJECT = "subject";
	private static String OBJECT = "object";
	private static String SUBJECT_TOKEN_INDICES = "subject_token_indices";
	private static String OBJECT_TOKEN_INDICES = "object_token_indices";

	public JSONObject toJSON() {
		return new JSONObject()
			.put(RELATION, relation)
			.put(POSITIVE, positive)
			.put(SENTENCE, sentence.toJSON())
			.put(SUBJECT, subject)
			.put(OBJECT, object)
			.put(SUBJECT_TOKEN_INDICES, JSONUtils.toIntArray(subjectTokenIndices))
			.put(OBJECT_TOKEN_INDICES, JSONUtils.toIntArray(objectTokenIndices));
	}

	public static TrainingSample fromJSON(JSONObject o) {
		TrainingSample sample = new TrainingSample();
		sample.relation = (String) o.get(RELATION);
		sample.positive = o.getBoolean(POSITIVE);
		sample.sentence = TrainingSampleParsedSentence.fromJSON((JSONObject) o.get(SENTENCE));
		sample.subject = (String) o.get(SUBJECT);
		sample.object = (String) o.get(OBJECT);
		sample.subjectTokenIndices = JSONUtils.arrayToIntList((JSONArray) o.get(SUBJECT_TOKEN_INDICES));
		sample.objectTokenIndices = JSONUtils.arrayToIntList((JSONArray) o.get(OBJECT_TOKEN_INDICES));
		return sample;
	}

	public static class TrainingSampleParsedSentence {
		public String text;
		public List<TrainingSampleSentenceToken> tokens;
		public String originArticle;
		public int originSentenceId;

		private static String TEXT = "text";
		private static String TOKENS = "tokens";
		private static String ORIGIN_ARTICLE = "origin_article";
		private static String ORIGIN_SENTENCE_ID = "origin_sentence_id";

		public JSONObject toJSON() {
			return new JSONObject()
				.put(TEXT, text)
				.put(TOKENS, JSONUtils.toArray(tokens, TrainingSampleSentenceToken::toJSON))
				.put(ORIGIN_ARTICLE, originArticle)
				.put(ORIGIN_SENTENCE_ID, originSentenceId);
		}

		public static TrainingSampleParsedSentence fromJSON(JSONObject o) {
			TrainingSampleParsedSentence s = new TrainingSampleParsedSentence();
			s.text = (String) o.get(TEXT);
			s.tokens = JSONUtils.arrayMap((JSONArray) o.get(TOKENS), TrainingSampleSentenceToken::fromJSON);
			s.originArticle = (String) o.get(ORIGIN_ARTICLE);
			s.originSentenceId = o.getInt(ORIGIN_SENTENCE_ID);
			return s;
		}
	}

	public static class TrainingSampleSentenceToken {
		public int startOffset;
		public int endOffset;
		public String lemma;
		public String pos;
		public String ner;

		private static String START_OFFSET = "start_offset";
		private static String END_OFFSET = "end_offset";
		private static String LEMMA = "lemma";
		private static String POS = "pos";
		private static String NER = "ner";

		public JSONObject toJSON() {
			return new JSONObject()
				.put(START_OFFSET, startOffset)
				.put(END_OFFSET, endOffset)
				.put(LEMMA, lemma)
				.put(POS, pos)
				.put(NER, ner);
		}

		public static TrainingSampleSentenceToken fromJSON(JSONObject o) {
			TrainingSampleSentenceToken t = new TrainingSampleSentenceToken();
			t.startOffset = o.getInt(START_OFFSET);
			t.endOffset = o.getInt(END_OFFSET);
			t.lemma = (String) o.get(LEMMA);
			t.pos = (String) o.get(POS);
			t.ner = (String) o.get(NER);
			return t;
		}
	}
}

File: src/prototype/lib/training_sample.py
from recordclass import recordclass

class TrainingSample(recordclass("TrainingSample",
                                 ["relation", "positive", "subject", "object",
                                  "sentence", "subject_token_indices",
                                  "object_token_indices"])):
    RELATION = 'relation'
    POSITIVE = 'positive'
    SENTENCE = 'sentence'
    SUBJECT = 'subject'
    OBJECT = 'object'
    SUBJECT_TOKEN_INDICES = 'subject_token_indices'
    OBJECT_TOKEN_INDICES = 'object_token_indices'

    def to_json(self):
        assert len(self.subject_token_indices) > 0
        assert len(self.object_token_indices) > 0

        return {
            self.RELATION: self.relation,
            self.POSITIVE: self.positive,
            self.SENTENCE: self.sentence.to_json(),
            self.SUBJECT: self.subject,
            self.OBJECT: self.object,
            self.SUBJECT_TOKEN_INDICES: self.subject_token_indices,
            self.OBJECT_TOKEN_INDICES: self.object_token_indices,
        }

    @classmethod
    def from_json(klass, json):
        self = klass(
            relation=json[klass.RELATION],
            positive=json[klass.POSITIVE],
            sentence=TrainingSampleParsedSentence.from_json(json[klass.SENTENCE]),
            subject=json[klass.SUBJECT],
            object=json[klass.OBJECT],
            subject_token_indices=json[klass.SUBJECT_TOKEN_INDICES],
            object_token_indices=json[klass.OBJECT_TOKEN_INDICES],
        )

        # Check that token indices look alright.
        assert len(self.subject_token_indices) > 0
        for index in self.subject_token_indices:
            if index not in range(len(self.sentence.tokens)):
                raise

        assert len(self.object_token_indices) > 0
        for index in self.object_token_indices:
            if index not in range(len(self.sentence.tokens)):
                raise

        return self

class TrainingSampleParsedSentence(recordclass("TrainingSampleParsedSentence",
                                               ["text", "tokens",
                                               "origin_article",
                                               "origin_sentence_id"])):
    TEXT = 'text'
    TOKENS = 'tokens'
    ORIGIN_ARTICLE = 'origin_article'
    ORIGIN_SENTENCE_ID = 'origin_sentence_id'

    def to_json(self):
        assert self.origin_article is not None

        return {
            self.TEXT: self.text,
            self.TOKENS: [token.to_json() for token in self.tokens],
            self.ORIGIN_ARTICLE: self.origin_article,
            self.ORIGIN_SENTENCE_ID: self.origin_sentence_id
        }

    @classmethod
    def from_json(klass, json):
        return klass(
            text = json[klass.TEXT],
            tokens = list(map(TrainingSampleSentenceToken.from_json,
                              json[klass.TOKENS])),
            origin_article = json[klass.ORIGIN_ARTICLE],
            origin_sentence_id = json[klass.ORIGIN_SENTENCE_ID],
        )

class TrainingSampleSentenceToken(recordclass("TrainingSampleSentenceToken",
                                              ["start_offset", "end_offset",
                                               "lemma", "pos", "ner"])):
    START_OFFSET = 'start_offset'
    END_OFFSET = 'end_offset'
    LEMMA = 'lemma'
    POS = 'pos'
    NER = 'ner'

    def to_json(self):
        return {
            self.START_OFFSET: self.start_offset,
            self.END_OFFSET: self.end_offset,
            self.LEMMA: self.lemma,
            self.POS: self.pos,
            self.NER: self.ner,
        }

    @classmethod
    def from_json(klass, json):
        return klass(
            start_offset = json[klass.START_OFFSET],
            end_offset = json[klass.END_OFFSET],
            lemma = json[klass.LEMMA],
            pos = json[klass.POS],
            ner = json[klass.NER],
        )
